compiler/fir/resolve/src/org/jetbrains/kotlin/fir/types/ConeTypeCompatibilityChecker.kt

```diff
@@ -0,0 +1,410 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.fir.types
+
+import org.jetbrains.kotlin.config.LanguageFeature
+import org.jetbrains.kotlin.descriptors.ClassKind
+import org.jetbrains.kotlin.descriptors.Modality
+import org.jetbrains.kotlin.fir.declarations.*
+import org.jetbrains.kotlin.fir.languageVersionSettings
+import org.jetbrains.kotlin.fir.resolve.getSymbolByLookupTag
+import org.jetbrains.kotlin.fir.resolve.symbolProvider
+import org.jetbrains.kotlin.fir.symbols.ConeClassLikeLookupTag
+import org.jetbrains.kotlin.fir.symbols.ConeTypeParameterLookupTag
+import org.jetbrains.kotlin.fir.symbols.StandardClassIds
+import org.jetbrains.kotlin.fir.symbols.impl.FirTypeParameterSymbol
+import org.jetbrains.kotlin.types.Variance
+import org.jetbrains.kotlin.utils.addToStdlib.cast
+import java.util.concurrent.atomic.AtomicInteger
+
+/**
+ * Checks if a given collection of [ConeKotlinType] are compatible. In other words, the types are compatible if it's possible at all to
+ * define a type that's a subtype of all of the given types. The compatibility of a given set of types concept is closely related to whether
+ * the intersection of these types is inhabited. But it's not identical because 1) one can manually control visibility of constructors and
+ * 2) there can be unused type parameters.
+ *
+ * The checker has two modes, which are controlled [LanguageFeature.ExtendedTypeCompatibilityChecker]:
```

Comment:
Do we even _want_ an FE1.0 compatible mode? Is there value in mimicking FE1.0 behavior (which seems incorrect in some places).

Reply:
Well I would be happy to remove this compatible mode. I am including it in the PR so that we don't have to reject code that was considered valid by FE1.0 if needed. It seems to be the way to introduce "breaking" changes so far. I can even add a warning diagnostic to smooth the transition if desired.